File: src/core/data_formats/parsing/ninja/motion.ts
import { ANGLE_TO_RAD } from "./index";
import { Cursor } from "../../block/cursor/Cursor";
import { Vec3 } from "../../vector";

const NMDM = 0x4d444d4e;

export type NjMotion = {
    motion_data: NjMotionData[];
    frame_count: number;
    type: number;
    interpolation: NjInterpolation;
    element_count: number;
};

export enum NjInterpolation {
    Linear,
    Spline,
    UserFunction,
}

export type NjMotionData = {
    tracks: NjKeyframeTrack[];
};

export enum NjKeyframeTrackType {
    Position,
    Rotation,
    Scale,
}

export type NjKeyframeTrack =
    | NjKeyframeTrackPosition
    | NjKeyframeTrackRotation
    | NjKeyframeTrackScale;

export type NjKeyframeTrackPosition = {
    type: NjKeyframeTrackType.Position;
    keyframes: NjKeyframeF[];
};

export type NjKeyframeTrackRotation = {
    type: NjKeyframeTrackType.Rotation;
    keyframes: NjKeyframeA[];
};

export type NjKeyframeTrackScale = {
    type: NjKeyframeTrackType.Scale;
    keyframes: NjKeyframeF[];
};

export type NjKeyframe = NjKeyframeF | NjKeyframeA;

/**
 * Used for parallel motion (POS), scale (SCL) and vector (VEC).
 */
export type NjKeyframeF = {
    frame: number;
    value: Vec3;
};

/**
 * Used for rotation (ANG).
 */
export type NjKeyframeA = {
    frame: number;
    value: Vec3; // Euler angles in radians.
};

export function parse_njm(cursor: Cursor, bone_count: number): NjMotion {
    if (cursor.u32() === NMDM) {
        return parse_njm_v2(cursor, bone_count);
    } else {
        cursor.seek_start(0);
        return parse_njm_bb(cursor, bone_count);
    }
}

/**
 * Format used by PSO v2 and for the enemies in PSO:BB.
 */
function parse_njm_v2(cursor: Cursor, bone_count: number): NjMotion {
    const chunk_size = cursor.u32();
    return parse_motion(cursor.take(chunk_size), bone_count);
}

/**
 * Format used by PSO:BB plymotiondata.rlc.
 */
function parse_njm_bb(cursor: Cursor, bone_count: number): NjMotion {
    cursor.seek_end(16);
    const offset1 = cursor.u32();
    cursor.seek_start(offset1);
    const action_offset = cursor.u32();
    cursor.seek_start(action_offset);
    return parse_action(cursor, bone_count);
}

function parse_action(cursor: Cursor, bone_count: number): NjMotion {
    cursor.seek(4); // Object pointer placeholder.
    const motion_offset = cursor.u32();
    cursor.seek_start(motion_offset);
    return parse_motion(cursor, bone_count);
}

function parse_motion(cursor: Cursor, bone_count: number): NjMotion {
    // Points to an array the size of bone_count.
    let mdata_offset = cursor.u32();
    const frame_count = cursor.u32();
    const type = cursor.u16();
    const inp_fn = cursor.u16();
    // Linear, spline or user function.
    const interpolation: NjInterpolation = (inp_fn & 0b11000000) >> 6;
    const element_count = inp_fn & 0b1111;
    const motion_data_list = [];

    for (let i = 0; i < bone_count; i++) {
        cursor.seek_start(mdata_offset);
        mdata_offset = mdata_offset += 8 * element_count;

        const motion_data: NjMotionData = {
            tracks: [],
        };
        motion_data_list.push(motion_data);

        const keyframe_offsets: number[] = [];
        const keyframe_counts: number[] = [];

        for (let j = 0; j < element_count; j++) {
            keyframe_offsets.push(cursor.u32());
        }

        for (let j = 0; j < element_count; j++) {
            const count = cursor.u32();
            keyframe_counts.push(count);
        }

        // NJD_MTYPE_POS_0
        if (type & (1 << 0)) {
            cursor.seek_start(keyframe_offsets.shift()!);
            const count = keyframe_counts.shift();

            if (count) {
                motion_data.tracks.push({
                    type: NjKeyframeTrackType.Position,
                    keyframes: parse_motion_data_f(cursor, count),
                });
            }
        }

        // NJD_MTYPE_ANG_1
        if (type & (1 << 1)) {
            cursor.seek_start(keyframe_offsets.shift()!);
            const count = keyframe_counts.shift();

            if (count) {
                motion_data.tracks.push({
                    type: NjKeyframeTrackType.Rotation,
                    keyframes: parse_motion_data_a(cursor, count, frame_count),
                });
            }
        }

        // NJD_MTYPE_SCL_2
        if (type & (1 << 2)) {
            cursor.seek_start(keyframe_offsets.shift()!);
            const count = keyframe_counts.shift();

            if (count) {
                motion_data.tracks.push({
                    type: NjKeyframeTrackType.Scale,
                    keyframes: parse_motion_data_f(cursor, count),
                });
            }
        }

        // TODO: all NJD_MTYPE's
    }

    return {
        motion_data: motion_data_list,
        frame_count,
        type,
        interpolation,
        element_count,
    };
}

function parse_motion_data_f(cursor: Cursor, count: number): NjKeyframeF[] {
    const frames: NjKeyframeF[] = [];

    for (let i = 0; i < count; ++i) {
        frames.push({
            frame: cursor.u32(),
            value: cursor.vec3_f32(),
        });
    }

    return frames;
}

function parse_motion_data_a(
    cursor: Cursor,
    keyframe_count: number,
    frame_count: number,
): NjKeyframeA[] {
    const frames: NjKeyframeA[] = [];
    const start_pos = cursor.position;

    for (let i = 0; i < keyframe_count; ++i) {
        frames.push({
            frame: cursor.u16(),
            value: {
                x: cursor.u16() * ANGLE_TO_RAD,
                y: cursor.u16() * ANGLE_TO_RAD,
                z: cursor.u16() * ANGLE_TO_RAD,
            },
        });
    }

    let prev = -1;

    for (const { frame } of frames) {
        if (frame < prev || frame >= frame_count) {
            cursor.seek_start(start_pos);
            return parse_motion_data_a_wide(cursor, keyframe_count);
        }

        prev = frame;
    }

    return frames;
}

function parse_motion_data_a_wide(cursor: Cursor, keyframe_count: number): NjKeyframeA[] {
    const frames: NjKeyframeA[] = [];

    for (let i = 0; i < keyframe_count; ++i) {
        frames.push({
            frame: cursor.u32(),
            value: {
                x: cursor.i32() * ANGLE_TO_RAD,
                y: cursor.i32() * ANGLE_TO_RAD,
                z: cursor.i32() * ANGLE_TO_RAD,
            },
        });
    }

    return frames;
}

File: lib/src/commonMain/kotlin/world/phantasmal/lib/fileFormats/ninja/Motion.kt
package world.phantasmal.lib.fileFormats.ninja

import world.phantasmal.core.isBitSet
import world.phantasmal.lib.cursor.Cursor
import world.phantasmal.lib.fileFormats.Vec3
import world.phantasmal.lib.fileFormats.vec3Float

private const val NMDM = 0x4d444d4e

class NjMotion(
    val motionData: List<NjMotionData>,
    val frameCount: Int,
    val type: Int,
    val interpolation: NjInterpolation,
    val elementCount: Int,
)

enum class NjInterpolation {
    Linear,
    Spline,
    UserFunction,
}

class NjMotionData(
    val tracks: List<NjKeyframeTrack>,
)

sealed class NjKeyframeTrack {
    abstract val keyframes: List<NjKeyframe>

    class Position(
        override val keyframes: List<NjKeyframe.Vector>,
    ) : NjKeyframeTrack()

    class EulerAngles(
        override val keyframes: List<NjKeyframe.Vector>,
    ) : NjKeyframeTrack()

    class Scale(
        override val keyframes: List<NjKeyframe.Vector>,
    ) : NjKeyframeTrack()

    class Quaternion(
        override val keyframes: List<NjKeyframe.Quaternion>,
    ) : NjKeyframeTrack()
}

sealed class NjKeyframe {
    abstract val frame: Int

    /**
     * Used for parallel motion (POS), scale (SCL), vector (VEC) or rotation (ANG).
     */
    class Vector(
        override val frame: Int,
        /**
         * Position, scale, vector, or euler angles in radians.
         */
        val value: Vec3,
    ) : NjKeyframe()

    class Quaternion(
        override val frame: Int,
        val real: Float,
        val imaginary: Vec3,
    ) : NjKeyframe()
}

fun parseNjm(cursor: Cursor): NjMotion =
    if (cursor.int() == NMDM) {
        parseNjmV2(cursor)
    } else {
        cursor.seekStart(0)
        parseNjmBb(cursor)
    }

/**
 * Format used by PSO v2 and for the enemies in PSO:BB.
 */
private fun parseNjmV2(cursor: Cursor): NjMotion {
    val chunkSize = cursor.int()
    return parseMotion(cursor.take(chunkSize), v2Format = true)
}

/**
 * Format used by PSO:BB plymotiondata.rlc.
 */
private fun parseNjmBb(cursor: Cursor): NjMotion {
    cursor.seekEnd(16)
    val offset1 = cursor.int()
    cursor.seekStart(offset1)
    val actionOffset = cursor.int()
    cursor.seekStart(actionOffset)
    return parseAction(cursor)
}

private fun parseAction(cursor: Cursor): NjMotion {
    cursor.seek(4) // Object pointer placeholder.
    val motionOffset = cursor.int()
    cursor.seekStart(motionOffset)
    return parseMotion(cursor, v2Format = false)
}

fun parseMotion(cursor: Cursor, v2Format: Boolean): NjMotion {
    // For v2, try to determine the end of the mData offset table by finding the lowest mDataOffset
    // value. This is usually the value that the first mDataOffset points to. This value is assumed
    // to be the end of the mDataOffset table.
    var mDataTableEnd = if (v2Format) cursor.size else cursor.position

    // Points to an array with an element per bone.
    val mDataTableOffset = cursor.int()
    val frameCount = cursor.int()
    val type = cursor.uShort().toInt()
    val inpFn = cursor.uShort().toInt()
    // Linear, spline or user private fun.
    val interpolation = when (val inp = (inpFn and 0b11000000) ushr 6) {
        0 -> NjInterpolation.Linear
        1 -> NjInterpolation.Spline
        // Default to UserFunction.
        2, 3 -> NjInterpolation.UserFunction
        else -> error("Interpolation bits should be between 0 and 3, inclusive but were ${inp}.")
    }
    val elementCount = inpFn and 0b1111
    val motionDataList = mutableListOf<NjMotionData>()
    var mDataOffset = mDataTableOffset

    while (mDataOffset < mDataTableEnd) {
        cursor.seekStart(mDataOffset)
        mDataOffset += 8 * elementCount

        val tracks = mutableListOf<NjKeyframeTrack>()

        val keyframeOffsets = mutableListOf<Int>()
        val keyframeCounts = mutableListOf<Int>()

        repeat(elementCount) {
            val keyframeOffset = cursor.int()
            keyframeOffsets.add(keyframeOffset)

            if (v2Format && keyframeOffset != 0 && keyframeOffset < mDataTableEnd) {
                mDataTableEnd = keyframeOffset
            }
        }

        repeat(elementCount) {
            keyframeCounts.add(cursor.int())
        }

        // NJD_MTYPE_POS_0
        if (type.isBitSet(0)) {
            cursor.seekStart(keyframeOffsets.removeFirst())
            val count = keyframeCounts.removeFirst()

            tracks.add(NjKeyframeTrack.Position(
                keyframes = parseVectorKeyframes(cursor, count),
            ))
        }

        // NJD_MTYPE_ANG_1
        if (type.isBitSet(1)) {
            cursor.seekStart(keyframeOffsets.removeFirst())
            val count = keyframeCounts.removeFirst()

            tracks.add(NjKeyframeTrack.EulerAngles(
                keyframes = parseEulerAngleKeyframes(cursor, count, frameCount),
            ))
        }

        // NJD_MTYPE_SCL_2
        if (type.isBitSet(2)) {
            cursor.seekStart(keyframeOffsets.removeFirst())
            val count = keyframeCounts.removeFirst()

            tracks.add(NjKeyframeTrack.Scale(
                keyframes = parseVectorKeyframes(cursor, count),
            ))
        }

        // NJD_MTYPE_QUAT_1
        if (type.isBitSet(13)) {
            cursor.seekStart(keyframeOffsets.removeFirst())
            val count = keyframeCounts.removeFirst()

            tracks.add(NjKeyframeTrack.Quaternion(
                keyframes = parseQuaternionKeyframes(cursor, count),
            ))
        }

        // TODO: all NJD_MTYPE's

        motionDataList.add(NjMotionData(tracks))
    }

    return NjMotion(
        motionData = motionDataList,
        frameCount,
        type,
        interpolation,
        elementCount,
    )
}

private fun parseVectorKeyframes(cursor: Cursor, count: Int): List<NjKeyframe.Vector> {
    val frames = mutableListOf<NjKeyframe.Vector>()

    repeat(count) {
        frames.add(NjKeyframe.Vector(
            frame = cursor.int(),
            value = cursor.vec3Float(),
        ))
    }

    return frames
}

private fun parseEulerAngleKeyframes(
    cursor: Cursor,
    keyframeCount: Int,
    frameCount: Int,
): List<NjKeyframe.Vector> {
    val frames = mutableListOf<NjKeyframe.Vector>()
    val startPos = cursor.position

    repeat(keyframeCount) {
        frames.add(NjKeyframe.Vector(
            frame = cursor.uShort().toInt(),
            value = Vec3(
                angleToRad(cursor.uShort().toInt()),
                angleToRad(cursor.uShort().toInt()),
                angleToRad(cursor.uShort().toInt()),
            ),
        ))
    }

    var prev = -1

    for (keyframe in frames) {
        if (keyframe.frame < prev || keyframe.frame >= frameCount) {
            cursor.seekStart(startPos)
            return parseEulerAngleKeyframesWide(cursor, keyframeCount)
        }

        prev = keyframe.frame
    }

    return frames
}

private fun parseEulerAngleKeyframesWide(
    cursor: Cursor,
    keyframeCount: Int,
): List<NjKeyframe.Vector> {
    val frames = mutableListOf<NjKeyframe.Vector>()

    repeat(keyframeCount) {
        frames.add(NjKeyframe.Vector(
            frame = cursor.int(),
            value = Vec3(
                angleToRad(cursor.int()),
                angleToRad(cursor.int()),
                angleToRad(cursor.int()),
            ),
        ))
    }

    return frames
}

private fun parseQuaternionKeyframes(
    cursor: Cursor,
    keyframeCount: Int,
): List<NjKeyframe.Quaternion> {
    val frames = mutableListOf<NjKeyframe.Quaternion>()

    repeat(keyframeCount) {
        frames.add(NjKeyframe.Quaternion(
            frame = cursor.int(),
            real = cursor.float(),
            imaginary = cursor.vec3Float(),
        ))
    }

    return frames
}
